src/Invitation.ts

```diff
@@ -0,0 +1,201 @@
+      throw new Error('Missing header bytes')
+    }
+    if (!ciphertext) {
+      throw new Error('Missing ciphertext')
+    }
+    this.headerBytes = headerBytes
+    this.ciphertext = new Ciphertext(ciphertext)
+  }
+
+  /**
+   * Accessor method for the full header object
+   */
+  get header(): SealedInvitationHeaderV1 {
+    // Use cached value if already exists
+    if (this._header) {
+      return this._header
+    }
+    this._header = SealedInvitationHeaderV1.fromBytes(this.headerBytes)
+    return this._header
+  }
+
+  /**
+   * getInvitation decrypts and returns the InvitationV1 stored in the ciphertext of the Sealed Invitation
+   */
+  async getInvitation(viewer: PrivateKeyBundleV2): Promise<InvitationV1> {
+    // Use cached value if already exists
+    if (this._invitation) {
+      return this._invitation
+    }
+    // The constructors for child classes will validate that this is complete
+    const header = this.header
+    let secret: Uint8Array
+    if (viewer.identityKey.matches(this.header.sender.identityKey)) {
+      secret = await viewer.sharedSecret(
+        header.recipient,
+        header.sender.preKey,
+        false
+      )
+    } else {
+      secret = await viewer.sharedSecret(
+        header.sender,
+        header.recipient.preKey,
+        true
+      )
+    }
+
+    const decryptedBytes = await decrypt(
+      this.ciphertext,
+      secret,
+      this.headerBytes
+    )
+    this._invitation = InvitationV1.fromBytes(decryptedBytes)
+    return this._invitation
+  }
+
+  toBytes(): Uint8Array {
+    return invitation.SealedInvitationV1.encode(this).finish()
+  }
+
+  static fromBytes(bytes: Uint8Array): SealedInvitationV1 {
+    return new SealedInvitationV1(invitation.SealedInvitationV1.decode(bytes))
+  }
+}
+
+/**
+ * Wrapper class for SealedInvitationV1 and any future iterations of SealedInvitation
+ */
+export class SealedInvitation implements invitation.SealedInvitation {
+  v1: SealedInvitationV1
+
+  constructor({ v1 }: invitation.SealedInvitation) {
+    if (!v1) {
+      throw new Error('Missing v1 invitation')
+    }
+    this.v1 = new SealedInvitationV1(v1)
+  }
+
+  toBytes(): Uint8Array {
+    return invitation.SealedInvitation.encode(this).finish()
+  }
+
+  static fromBytes(bytes: Uint8Array): SealedInvitation {
+    return new SealedInvitation(invitation.SealedInvitation.decode(bytes))
+  }
+
+  /**
+   * Create a SealedInvitation with a SealedInvitationV1 payload
+   * Will encrypt all contents and validate inputs
+   */
+  static async createV1({
+    sender,
+    recipient,
+    created,
+    invitation,
+  }: {
+    sender: PrivateKeyBundleV2
+    recipient: SignedPublicKeyBundle
+    created: Date
+    invitation: InvitationV1
+  }): Promise<SealedInvitation> {
+    const headerBytes = new SealedInvitationHeaderV1({
+      sender: sender.getPublicKeyBundle(),
+      recipient,
+      createdNs: Long.fromNumber(created.valueOf()).multiply(1_000_000),
```

Comment:
We should probably create a helper for this, `nsNow()` or something, and hunt down all the places we do these conversions.

Reply:
Agreed. I can fold that in